IPLD-Data-Model-v1.md

```diff
@@ -1,17 +1,103 @@
-# [WIP] IPLD Data Model
+IPLD Data Model
+===============
 
-## Simple Types
+The IPLD Data Model is "layer 1" of the IPLD core models.
+("Layer 0" are the [Codecs](Codecs); "layer 2" is called [Schemas](schema).)
+
+The Data Model specifies what categories of data are representable in IPLD --
+for example, booleans, and integers, and textual strings.
+[Codecs](Codecs) specify exactly how these data are transcribed into serialized bytes.
+
+
+Kinds
+-----
+
+The following is the list of essential _kinds_ (or more formally, _representation kinds_)
+of data representable in the IPLD Data Model:
 
-* Boolean
 * Null
+* Boolean
+* Integer
+* Float
 * String
+* Bytes
+* List
+* Map
+* Link
+
+(Note that we use the term "_kinds_" here to disambiguate this from "_types_",
+which is a term we'll use at the [Schemas](schema) level.)
+
+The _recursive kinds_ are:
+
+* List
+* Map
+
+The _scalar kinds_ (the complement of recursive) are:
+
+* Null
+* Boolean
 * Integer
 * Float
-* Array
-* Object (Hash Map)
-* Binary
+* String
+* Bytes
+* Link
+
+(Note that [Schemas](schema) introduce a few more kinds -- when clarification is necessary,
+these Data Model kinds can be called the "_representation kinds_",
+while the additional kinds introduced in the Schema layer are "_perceived kinds_".)
+
+
+### Null kind
+
+Null is a scalar kind.  Its cardinality is one -- the only value is 'null'.
+
+### Boolean kind
 
-## Link Type
+Boolean is a scalar kind.  Its cardinality is two -- either the value 'true' or the value 'false'.
 
-This type represents a link to another IPLD Block. The link reference
+### Integer kind
+
+
+### Float kind
+
+
+### String kind
+
+String is a scalar kind.  Its cardinality is infinite -- strings do not have a length limit.
+
+Strings are generally understood to be UTF-8 text, in [NFC](https://www.unicode.org/reports/tr15/#Norm_Forms) canonicalization.
```

Comment:
n.b. I'm stating this one as a fact, because afaik, the last time string normalization was discussed was in https://github.com/ipld/specs/issues/67, and there seems to be consensus that NFC is the most widely accepted answer.

I also find the Golang approach of "strings are just bytes, actually; glhf" to be quite practical, which is why I used the somewhat loose phrasing of "generally understood".  We might want to lock things down even more explicitly when it comes to map keys (for pathability reasons), but that's a bigger issue than we probably want to engage with in the scope of this PR.

Reply:
I would change "generally considered" to the RFC "MUST" term. I don't think we can leave any ambiguity here.

Reply:
I'll go for "should".  In lowercase, but correct to interpret in RFC 2119 terms.

I've noticed some other modern specs, even when using RFC 2119 terms (and explicitly denoting it in their preambles) are dropping the capslock on the terms.  If we do use RFC 2119, I'd vote for doing the same.  The capslock adds a lot of "line noise", IMO; and I've been reading the specs that drop the caps without any confusion.

We might upgrade this to "must" at some point, but I'd rather give that it's own discussion space.  Ideally some concrete ideas about programmatically how we'd enforce that, some intuitions about the performance impact of it, and most importantly how we'd recommend raising the errors would all be good to have in hand before going into that.